### ISSUE TITLE
Explore ways to improve retriever accuracy for Table+Text

### ISSUE THREAD

**Issue by jdpsen** [primeqa/primeqa#21]
_jaydesen created the following on Jan 25:_  

 We need to explore ways and test different approaches on how to better the retrieval accuracies at different levels: Table retrieval, row retrieval and linked passage retrieval.  

_Original issue: https://github.ibm.com/ai-foundation/PrimeQA/issues/12_

**Comment by jdpsen** [comment]
_On Feb 01, vishk024 commented:_
- We observed that there is significant drop in final EM score due to incorrect row-passage linking and incorrect table retrieval.
####  In order to overcome these issues with table retrieval we plan to do following things:

1. Use standard entity linker such as GENRE by Facebook research instead of current GPT-2+BM25 row passage linker.
2. Experiment with Distantly supervised iterative retriever [DistDR](https://arxiv.org/abs/2110.04889) to improve table and passage retrieval accuracy.
3. We plan to use passage retrieval score to improve table retrieval and table retriever score to improve passage retrieval in an iterative fashion (Which we didn't try till now) table retrieval and passage retrieval are independent.

- We observed that row-retriever accuracy could be improved by incorporating table title and caption, we plan to train the row retriever using table title and caption and we are hoping to get a further boost in performance.

- We didn't train our answer extractor (AE) module using ott-qa train dataset(for SIGIR 2022 submission), we hope to get 1-2% boost in final answer EM by training the AE using ott-qa train set.

- We plan to use some additional set of features while re-ranking the final answer, Currently we just use row-retrieval confidence, answer start index probability and answer end-index probability to rank the final answers.

**Comment by jdpsen** [comment]
_On Feb 07, avi commented:_
We should also explore our ColBERT approach. Maybe @franzm @yulongl can discuss with @vishk024 @saneem-cg ?

**Comment by jdpsen** [comment]
_On Feb 08, jaydesen commented:_
One idea is : We can try query token aware iterative retrieval using colBERT instead of DPR.

**Comment by jdpsen** [comment]
_On Feb 08, franzm commented:_
Ad "query token aware iterative retrieval using colBERT":
Using a "compressed" version of ColBERT representation, i.e. query represented by fewer than 32 vectors as in original ColBERT by more than one vector as in DPR is definitely a possibility.

**Comment by jdpsen** [comment]
_On Feb 15, vishk024 commented:_
Currently we are trying to improve recall of passage ranking using following approach:

- Generate text using a table row using pre-trained GPT-2
- Using Pre-trained GENRE system to get a list of wiki links for the entities in the table row.

For a given table: 
|year|prize|Winner|
|1921|Nobel Prize | Einstein|

GPT2 Generates: `In 1921, Einstein received a Nobel Prize.`
GENRE Entity linker outputs:
```
[[{'text': 'In 1921, { Einstein } [ Albert Einstein ] received a { Nobel Prize } [ Nobel Prize in Physiology or Medicine ].', 'logprob': tensor(-0.9068)}], [{'text': 'In 1921, { Einstein } [ Albert Einstein ] received a { Nobel Prize } [ Nobel Prize in Physiology or Physiology ].', 'logprob': tensor(-1.0778)}], [{'text': 'In 1921, { Einstein } [ Albert Einstein ] received a { Nobel Prize } [ Nobel Prize in Physiology or Medicine ] {. } [ Ernest Einstein ]', 'logprob': tensor(-1.1164)}], [{'text': 'In 1921, { Einstein } [ Albert Einstein ] received a { Nobel Prize } [ Nobel Prize in Physiology or Medicine ] {. } [ Einstein ]', 'logprob': tensor(-1.1661)}], [{'text': 'In 1921, { Einstein } [ Albert Einstein ] received a { Nobel Prize } [ Nobel Prize in Physiology or Medicine ] {. } [ Max Einstein ]', 'logprob': tensor(-1.1712)}]]
```

**Comment by jdpsen** [comment]
_On Mar 01, vishk024 commented:_
The GENRE system expects a perfect contextual sentence as input. We attempted to use it using "is" and "." as separators between column header and columns. For example the table row is converted to this format: `Rank is 1 . Player is Emmitt Smith . Team ( s ) by season is Dallas Cowboys ( 1990 - 2002 ) Arizona Cardinals ( 2003 - 2004 ) . Carries is 4,409 . Yards is 18,355 . Averge is 4.2 . `. As the entity linking technique used in GENRE is autoregressive we observed that the links annotate by GENRE system is sometimes does not even have a wikipedia page. Also, it s bit tricky to use the pre-trained GENRE model and it removes some blank spaces, making it difficult to do entity mention matching. So, we decided to drop the idea of using GENRE as entity linker to increase recall.

#### Question to passage retrieval
Next we attempted to use DPR to retrieve passages based on question representation independently and append the links thus retrieved to all existing links for every row.
We have some initial numbers using this approach and it seems to further drop the performance a bit. The dev set performance is: 
`OrderedDict([('total exact', 25.609756097560975), ('total f1', 30.312986346184175), ('total', 2214)])` The best number reported on dev set in the paper is `EM: 40.0 and  F1:  45.1`

**Comment by jdpsen** [comment]
_On Mar 08, jaydesen commented:_
GENRE system fails. 
Simply adding DRP retrieved passages to each row does not work. 
manually verifying q->p shows the potential of better recall.

DistDR- setup is done and trials to start. DistDR performs better than MDR

**Comment by jdpsen** [comment]
_On Mar 15, vishk024 commented:_
Q->P matching does give some passages relevant to the query but it affect passage filtering performance and Q-> P passages dominates over T -> P that ultimately affects the final answer prediction performance. We get a result very similar to previous numbers i.e 25% EM and 30.00% F1. 

Now we are going to try out Colbert for Q -> T matching and Q-> P matching and hope that this will increase the overall performance.

DistDR code is working fine for NQ now we are going to split 400K tables to 4M tables, linearise and train DistDR for OTTQA.

**Comment by jdpsen** [comment]
_On Mar 24, jaydesen commented:_
@vishk024 : update the distDR results

**Comment by jdpsen** [comment]
@vishwajeetkumar93 : please update the current status of retriever imporvements.

**Comment by jdpsen** [comment]
This aligns with #130 and should be tracked by that
